### PR TITLE
Offline first capabilities

### PR DESCRIPTION
**Pull request by chege4179**
This pull request introduces offline first capabilities to the dcke 23 app by using work manager ro sync the local data source(Room) with the remote data source (Ktor client). By default the repos fetch data from the local data source to enable offline capabilites and are synced when the internet is back.I created abstractions for the local and remote data sources to help separate concerns. For now I did not alter much of the sessions Repo because of the bookmarks functionality but will do so in a future PR

### CODE REVIEW

**Comment by wangerekaharun** [comment]
@chege4179 check build failures first

**Comment by chege4179** [comment]
The workflow fails because of the dete linter
I am not sure on how to solve it

**Comment by chege4179** [comment]
@wangerekaharun I've fixed the failing CI

**Comment by wangerekaharun** [comment]
> @wangerekaharun I've fixed the failing CI

Great I'll have a look

File: data/src/main/java/com/android254/data/repos/local/LocalSessionsDataSourceImpl.kt [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    private val sessionDao: SessionDao,
+    @IoDispatcher private val ioDispatcher: CoroutineDispatcher
+) : LocalSessionsDataSource {
+    override fun getCachedSessions(): Flow<List<Session>> {
+        return sessionDao.fetchSessions()
+            .flowOn(ioDispatcher)
+            .map { sessions -> sessions.map { it.toDomainModel() } }
+    }
+
+    @RequiresApi(Build.VERSION_CODES.O)
```
Comment (wangerekaharun) [inline data/src/main/java/com/android254/data/repos/local/LocalSessionsDataSourceImpl.kt:62]: Renove this

File: data/src/main/java/com/android254/data/work/SyncDataWorker.kt [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                is ResourceResult.Success -> {
+                    localSponsorsDataSource.deleteCachedSponsors()
+                    localSponsorsDataSource.saveCachedSponsors(
+                        sponsors = response.data ?: emptyList()
+                    )
+                    println("Sync sponsors successful")
+                }
+
+                is ResourceResult.Error -> {
+                    println("Sync sponsors failed ${response.message}")
```
Comment (wangerekaharun) [inline data/src/main/java/com/android254/data/work/SyncDataWorker.kt:99]: Timber would be better

File: presentation/src/main/java/com/android254/presentation/speakers/view/SpeakersScreen.kt [hunk 3]
```diff
@@ -94,46 +90,23 @@ fun SpeakersScreen(
...
-                        contentPadding = PaddingValues(start = 16.dp, end = 16.dp, bottom = 16.dp)
-                    ) {
-                        items(items = uiState.speakers) { speaker ->
-                            SpeakerComponent(
-                                speaker = speaker,
-                                onClick = {
-                                    navigateToSpeaker.invoke(speaker.id)
-                                }
-                            )
+            LazyVerticalGrid(
```
Comment (wangerekaharun) [inline presentation/src/main/java/com/android254/presentation/speakers/view/SpeakersScreen.kt:99]: How will you propagate errors here? eh empty state, error state?

**Review by wangerekaharun — APPROVED** [review]
Looks good, left some minor comments

File: data/src/main/java/com/android254/data/repos/local/LocalSessionsDataSourceImpl.kt (hunk 1 above)
Comment (chege4179) [reply data/src/main/java/com/android254/data/repos/local/LocalSessionsDataSourceImpl.kt:62]: Noted

File: data/src/main/java/com/android254/data/work/SyncDataWorker.kt (hunk 2 above)
Comment (chege4179) [reply data/src/main/java/com/android254/data/work/SyncDataWorker.kt:99]: My bad ...I wanted to log something quickly
I will replace it

File: presentation/src/main/java/com/android254/presentation/speakers/view/SpeakersScreen.kt (hunk 3 above)
Comment (chege4179) [reply presentation/src/main/java/com/android254/presentation/speakers/view/SpeakersScreen.kt:99]: I will modify it

**Comment by wangerekaharun** [comment]
all good now @chege4179

**Comment by chege4179** [comment]
yeah ...was waiting for the other reviewers

File: data/src/main/java/com/android254/data/di/DataSourcesModule.kt [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+import dagger.Module
+import dagger.Provides
+import dagger.hilt.InstallIn
+import dagger.hilt.components.SingletonComponent
+import kotlinx.coroutines.CoroutineDispatcher
+import javax.inject.Singleton
+
+@Module
+@InstallIn(SingletonComponent::class)
+object DataSourcesModule {
```
Comment (chepsi) [inline data/src/main/java/com/android254/data/di/DataSourcesModule.kt:45]: minor `DataSourcesModule` -> `DataSourceModule`

File: data/src/main/java/com/android254/data/repos/local/LocalSessionsDataSourceImpl.kt [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    suspend fun fetchSessionWithFilters(query: SupportSQLiteQuery): List<Session>
+
+    suspend fun updateBookmarkedStatus(id: String, isBookmarked: Boolean)
+
+    suspend fun getBookmarkStatus(id: String): Boolean
+
+    suspend fun saveCachedSessions(sessions: List<SessionDTO>)
+}
+
+class LocalSessionsDataSourceImpl @Inject constructor(
```
Comment (chepsi) [inline data/src/main/java/com/android254/data/repos/local/LocalSessionsDataSourceImpl.kt:52]: can we extract this to a separate file?

**Review by chepsi — APPROVED** [review]

**Pull request merged**